nixos/services.dictd: init gcide at 0.54

S0AndS0 opened the pull request:
GCIDE AKA GNU version of the Collaborative International Dictionary of English is a locally host-able database compatible with `dictd` service.

This _should_ fix #248974 by yoinking, and translating, AUR's packaging of [`dict-gcide`](https://aur.archlinux.org/packages/dict-gcide) into something that conforms to NixOS expectations.

And may also address questions found on [discourse](https://discourse.nixos.org/t/dict-offline-version/33004)

:warning: `sed-scripts` are a source of future fragility, and I ain't happy about how `libmaa` dependency is currently being stolen into these proposed changes.

On a positive note; local tests such as `dict hello` return, almost instantly, results from the database

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done


- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

---

Edit 2025-07-22 I see I'm failing the Lint checks, and will get 'round to trying to sort that _soon_ :facepalm: 

Update 2025-07-22 the [`error: path './sed-scripts/' has a trailing slash`](https://github.com/NixOS/nixpkgs/actions/runs/16460696703/job/46527220793?pr=427615#step:5:209) is, uh, kinda strange and I'd appreciate a bit o' expert help on that one :pray:

dramforever reviewed (changes requested):
Thanks for your contribution! I only realized now that the comments below have reached double digits, but I hope you don't mind that these are what brings your contributions more in line with what Nixpkgs expects.

I haven't tried it out yet, and I don't  fullyunderstand how the build process works, so hopefully someone with more energy or familiarity can provide further insights.

In addition to code suggestions, the commit message for the second commit should be `dictdDBs.gcide: init at 0.53`. Also please remember to squash any fixups at the end, so the history remains clean.

dramforever commented on pkgs/servers/dict/gcide/default.nix:
```diff
@@ -0,0 +1,126 @@
+  };
+
+  ## TODO: make this not ugly?
+  libmaa = builtins.head (builtins.filter (x: x.pname == "libmaa") pkgs.dict.buildInputs);
```
> TLDR: this'll have to be a change just before swashing commits and, for reasons, libmaa = pkgs.callPackage ../libmaa.nix { }; in the meantime will be used as a less obnoxious way of obtaining dependency x-)

Huh, I didn't expect that.

If it's not too much of an ask, consider bringing out `libmaa` into `pkgs/by-name`, and possibly also updating it to the latest version?

dramforever replied:
It should be fine. Renames are handled well by git

S0AndS0 commented on pkgs/servers/dict/gcide/default.nix:
```diff
@@ -0,0 +1,126 @@
+  '';
+
+  meta = {
+    description = "GNU version of the Collaborative International Dictionary of English for dictd et al";
```
Kleptoed the `meta.description`, and just about everything else, from the [AUR `dict-gcide`](https://aur.archlinux.org/packages/dict-gcide) `PKGBUILD` script I did x-)

I'll be sure to update though to what GNU has instead!

dramforever commented:
I want to clarify before any misunderstanding that by squash, I mean eventually getting down to three commits: Adding yourself to maintainers list, extracting libmaa, dictdDBS.gcide: init

S0AndS0 commented:
> I want to clarify before any misunderstanding that by squash, I mean eventually getting down to three commits: Adding yourself to maintainers list, extracting libmaa, dictdDBS.gcide: init

Ah, okay that seems reasonable, and I'll get on re-ordering changes and squashing thangs

S0AndS0 committed:
maintainers: add S0AndS0

dramforever commented on pkgs/servers/dict/gcide/default.nix:
```diff
@@ -0,0 +1,111 @@
+      s/${_major_debver}/${version}/ ;
+    }" scan.l;
+
+    rm config.guess config.h.in config.sub configure install-sh;
```
Also, this should go in `prePatch`

S0AndS0 committed:
libmaa: init at 1.3.2

Extracted and version updated as part of #427615 change requests;

    https://github.com/NixOS/nixpkgs/pull/427615#discussion_r2231622802

Version `1.5.1` as of date on this commit, 2025-07-25, seems the latest
`libmaa` package update around 2024-11-16 on SourceForge;

    https://sourceforge.net/projects/dict/

...  However, that fails to build due to upstream Makefile missing a
separator on line 8 so this patch set is rename/move only.

Due diligence steps preformed to mitigate breaking `dict`/`dictd` package;

- Hunt for `libmaa` references;
   ```bash
   find {pkgs,nixos} -type f -iname '*.nix' -exec grep -nHE 'libmaa' {} \;
   ```
- Hunt for `dict(d)?` references;
   ```bash
   find {pkgs,nixos} -type f -iname '*.nix' -exec grep -nHE '\s?dictd?\s+=' {} \;
   ```

S0AndS0 committed:
dictdDBs.gcide: init at 0.54

This _should_ fix #248974 by yoinking, and translating, AUR's packaging
of [`dict-gcide`](https://aur.archlinux.org/packages/dict-gcide) into
something that conforms to NixOS expectations.

And may also address questions found on
[discourse](https://discourse.nixos.org/t/dict-offline-version/33004)

:warning: `sed-scripts` are a source of future fragility, and updates
here for NixOS will lag behind those provided by AUR team.

Big thanks to @respiranto for adding license info to sed scripts!

```bash
git clone https://aur.archlinux.org/dict-gcide.git &&
  git checkout ff81bcc526e3716f58a616a95d1f12b00ceb5695
```

S0AndS0 commented:
Thanks dramforever for sticking with me, and ensuring I don't make a total mess of things!

Latest force-push (commit hash `3b6c7f24a0b6`) _should_ include almost all change requests, and local build/tests seem to be okay :tada:

> `[root@nixos:/etc/nixos]# nixos-rebuild switch --impure --flake . && dict hello`

```
warning: Git tree '/etc/nixos' is dirty
building the system configuration...
warning: Git tree '/etc/nixos' is dirty
#> ...
Done. The new configuration is /nix/store/377vr3j69x4kbhmkgd777pabyh7l12kv-nixos-system-nixos-25.11.20250708.9807714
1 definition found

From The Collaborative International Dictionary of English v.0.53 [dict-gcide]:

  Hello \Hel*lo"\, interj. & n.
     An exclamation used as a greeting, to call attention, as an
     exclamation of surprise, or to encourage one. This variant of
     {Halloo} and {Holloo} has become the dominant form. In the
     United States, it is the most common greeting used in
     answering a telephone.
     [1913 Webster +PJC]
```

dramforever reviewed (commented):
Two other things:

- The sed scripts have unclear license. Shipping it in Nixpkgs would amount to relicensing it to MIT. We should ask the AUR author about it.
- In general, a `fooPhase` needs `runPhase preFoo` at the start and `runPhase postFoo` at the end. This is a limitation in the current phase framework of Nixpkgs, but it is what it is.

S0AndS0 commented:
> The sed scripts have unclear license. Shipping it in Nixpkgs would amount to relicensing it to MIT. We should ask the AUR author about it.

Smart!  I figured the "Package Details" → "Licenses" value was _the_ project's license, but also get how even though it's a copy-left license it could technically apply to the package being distributed only.

Consider it [asked](https://aur.archlinux.org/packages/dict-gcide#comment-1034016) :-)
And I got notifications turned-on, so if/when the team answers I'll be sure to 

> In general, a `fooPhase` needs `runPhase preFoo` at the start and `runPhase postFoo` at the end. This is a limitation in the current phase framework of Nixpkgs, but it is what it is.

:thinking: I'm not sure if this is something I need to take action on, because I moved stuff to the `prePatch` phase as requested, but if it be more of a keep it in mind for any future packaging then I certainly will.  And am grateful regardless!

> No `...` here please. Again, check how it's done on similar packages.

Ooof, okay that's a bit embarrassing :facepalm:
Here I thought I read the ellipsis was needed to prevent errors of too many args, but nope!  And re-looking at neighboring packages I should've been smorter

Latest force push will have that removed as well as missing semicolons at the `checkPhase` added.  Please do let me know if there's anything else I gotta tackle while with await an answer from AUR maintainers!

S0AndS0 commented:
> > > In general, a `fooPhase` needs `runPhase preFoo` at the start and `runPhase postFoo` at the end. This is a limitation in the current phase framework of Nixpkgs, but it is what it is.
> > 
> > 
> > 🤔 I'm not sure if this is something I need to take action on, because I moved stuff to the `prePatch` phase as requested, but if it be more of a keep it in mind for any future packaging then I certainly will. And am grateful regardless!
> 
> I should have spelled it out, but in any case I simply meant `installPhase` and `buildPhase`

Heh, thanks!  I can be kinda dense, sometimes, and will for sure spend some quality time with the docs before submitting any new PRs

respiranto commented:
I have added licensing information on the sedfiles and such in the AUR package.

S0AndS0 commented:
@respiranto you're the best!  I noticed in the `diff` some things were removed, other than comments, and a few executable lines were added too.  But rebuild against version `0.53` seems to be functional, and run-time is happy, so I'm gonna trust ya ;-)

@dramforever I tried bumping the version locally to `0.54` but that led to `fetchurl` time-outs so, for now, I'm gonna keep the proposed version at `0.53` for folks to enjoy and report back on.  Please do let me know, when ya got a chance, if anything else need be done by me to get this PR into an acceptable state.

respiranto commented:
@S0AndS0:

* You may have used an older version of the AUR package; my commits from today definitely only added licensing.  The previous commit is from 2025-07-14.
* The current state of the sedfiles is specifically adapted to version 0.54; it may work for 0.53, but I actually removed many no-longer-needed fixes on update for version 0.54.

S0AndS0 commented:
>     * You may have used an older version of the AUR package; my commits from today definitely only added licensing.  The previous commit is from 2025-07-14.
> 
>     * The current state of the sedfiles is specifically adapted to version 0.54; it may work for 0.53, but I actually removed many no-longer-needed fixes on update for version 0.54.

Ah thanks for clarifying where other changes be from, and I'll try bumping the version here tomorrow to `0.54` as it my have been a rate-limiting thing that caused time-outs :shrug:

S0AndS0 commented:
Hey all!  Is there anything left for me to do for this PR, or are we awaiting a release schedule sort of thing?

S0AndS0 commented:
> I don't have permissions to merge. Please try `#review-requests:nixos.org` on matrix or https://discourse.nixos.org/t/prs-ready-for-review/3032

:saluting_face: can and will do!...  And now has done over on Matrix :-)

sikmir commented on pkgs/servers/dict/gcide/default.nix:
```diff
@@ -0,0 +1,105 @@
+    fi
+  '';
+
+  installPhase = ''
+    install -Dm 0644 -t "$out/share/dictd/" ./gcide.{dict.dz,index};
+
+    echo "en_US.UTF-8" > "$out/share/dictd/locale";
+
+    install -Dm 0644 -t "$out/share/doc/dict-gcide/" ../gcide-${version}/{README,INFO,pronunc.txt};
+  '';
```
```suggestion
  installPhase = ''
    runHook preInstall
    install -Dm 0644 -t "$out/share/dictd/" ./gcide.{dict.dz,index};

    echo "en_US.UTF-8" > "$out/share/dictd/locale";

    install -Dm 0644 -t "$out/share/doc/dict-gcide/" ../gcide-${version}/{README,INFO,pronunc.txt};
    runHook postInstall
  '';
```

sikmir commented on pkgs/servers/dict/gcide/default.nix:
```diff
@@ -0,0 +1,105 @@
+
+  enableParallelBuilding = false;
+  buildPhase = ''
+    make -j1;
```
Do we really need `-j1` with `enableParallelBuilding = false;`?

sikmir commented on pkgs/servers/dict/gcide/default.nix:
```diff
@@ -0,0 +1,105 @@
+  sourceRoot = "./${pname}";
+  srcs = [
+    (fetchurl {
+      url = "https://deb.debian.org/debian/pool/main/d/${pname}/${pname}_${_debver}.tar.xz";
```
```suggestion
      url = "https://deb.debian.org/debian/pool/main/d/dict-gcide/dict-gcide_${_debver}.tar.xz";
```
Otherwise overriding `pname` will result to invalid url.

S0AndS0 replied:
Oof good catch!...  I'll have that fixed too _soon™_

> Side note; as I don't wanna clobber the diff you're currently reviewing I'll be awaiting a clear signal from ya before I force-push/squash changes

S0AndS0 commented:
:warning: Some-when between first opening this PR and Git ref `d0fc30899600b9b3466ddb260fd83deb486c32f1` from `nixos-unstable` there now seems to be a an issue with `dict` or `dictd` reporting;

> `gcide is not a valid database, use -D for a list`

SuperSandro2000 commented on pkgs/servers/dict/gcide/default.nix:
```diff
@@ -0,0 +1,106 @@
+    post_webfilter = ./sed-scripts/post_webfilter.sed;
+  };
+
+  _major_debver = "0.48";
+  _debver = "${_major_debver}.5+nmu4";
```
```suggestion
  major_debver = "0.48";
  debver = "${major_debver}.5+nmu4";
```
those are already private

SuperSandro2000 commented on pkgs/servers/dict/gcide/default.nix:
```diff
@@ -0,0 +1,106 @@
+    flex
+  ];
+
+  prePatch = ''
```
```suggestion
  postPatch = ''
```
otherwise we break patches

SuperSandro2000 commented on pkgs/servers/dict/gcide/default.nix:
```diff
@@ -0,0 +1,106 @@
+    }" scan.l;
+  '';
+
+  buildPhase = ''
+    make;
+
+    ## Do the conversion explicitly, instead of `make db', to account for all
```
```suggestion
  postBuild = ''
    ## Do the conversion explicitly, instead of `make db', to account for all
```
Or where does that fail?

SuperSandro2000 commented on pkgs/servers/dict/gcide/default.nix:
```diff
@@ -0,0 +1,106 @@
+      ./webfilter |
+      sed -Ef "${sed-scripts.post_webfilter}" |
+      tee pre_webfmt.data |
+      LANG=C ./webfmt -c;
```
```suggestion
      LANG=C ./webfmt -c
```
the ; are not required
